lib/de.ml

```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
     d.s <- Header ;
     d.k <- decode_k ;
     WInf.reset d.w
+
+  module Non_streamable = struct
+    type decoder =
+      { i: bigstring
+      ; mutable i_pos : int
+      ; mutable i_len : int
+      ; mutable hold : int
+      ; mutable bits : int
+      ; o : bigstring
+      ; mutable o_pos : int
+      ; w : WInf.t
+      ; mutable l : int (* literal / length *)
+      ; mutable d : int (* distance *)
+      }
+
+    (* errors. *)
+
+    type error =
+      | Unexpected_end_of_input
+      | Invalid_kind_of_block
+      | Invalid_dictionary
+      | Invalid_complement_of_length
+      | Invalid_distance
+      | Invalid_distance_code
+
+    let pp_error ppf e =
+      let s = match e with
+      | Unexpected_end_of_input      -> "Unexpected_end_of_input"
+      | Invalid_kind_of_block        -> "Invalid_kind_of_block"
+      | Invalid_dictionary           -> "Invalid_dictionary"
+      | Invalid_complement_of_length -> "Invalid_complement_of_length"
+      | Invalid_distance             -> "Invalid_distance"
+      | Invalid_distance_code        -> "Invalid_distance_code"
+      in
+      Format.fprintf ppf "%s" s
+
+    exception Malformed of error
+
+    let err_unexpected_end_of_input () =
+      raise (Malformed Unexpected_end_of_input)
+
+    let err_invalid_kind_of_block () =
+      raise (Malformed Invalid_kind_of_block)
+
+    let err_invalid_dictionary () =
+      raise (Malformed Invalid_dictionary)
+
+    let err_invalid_complement_of_length () =
+      raise (Malformed Invalid_complement_of_length)
+
+    let err_invalid_distance () =
+      raise (Malformed Invalid_distance)
+
+    let err_invalid_distance_code () =
+      raise (Malformed Invalid_distance_code)
+
+    (* remaining bytes to read [d.i]. *)
+    let i_rem d = d.i_len - d.i_pos
+    [@@inline]
+
+    let _fill_byte d =
+      if i_rem d <= 0
+      then
+        err_unexpected_end_of_input ();
+      d.hold <- d.hold lor ((unsafe_get_uint8 d.i d.i_pos) lsl d.bits);
+      d.i_pos <- d.i_pos + 1;
+      d.bits <- d.bits + 8
+    [@@inline]
+
+    let rec fill_bits d n =
+      if d.bits < n
+      then
+        ( _fill_byte d
+        ; fill_bits d n)
+
+    let pop_bits d n =
+      let v = d.hold land (1 lsl n - 1) in
+      d.hold <- d.hold lsr n ;
+      d.bits <- d.bits - n ;
+      v
+    [@@inline]
+
+    let flat d =
+      d.hold <- d.hold lsr (d.bits land 7) ;
+      d.bits <- d.bits land (lnot 7) ;
```

Comment:
Those lines align the hold on 8 bits.
It can be found in the streamable implementation [there](https://github.com/mirage/decompress/blob/e21bd8dbfc10dd35abd9947902c613f7ec870dac/lib/de.ml#L832)

Reply:
Ok, so the situation is a bit more clear for me. According to [RFC1951](https://tools.ietf.org/html/rfc1951#section-3.2.4), `len` and `nlen` are aligned. At least, the worst situation for you is when you have more than 8 bits in your `hold`. In that case, a part of `len` is into `hold`.

I think and in your situation, the best is to re-update `d.i` (move back) if you have strictly more than 7 bits in your `hold` and then, do two simple `unsafe_get_uint16` to get `len` and `nlen`.